### ISSUE TITLE
CachedImage.InvalidateCache has no effect

### ISSUE THREAD

**Issue by escamoteur** [molinch/FFImageLoading#147]
Hi,

I have a listview with CachedImages bound to an ObservableCollection. When I change elements of the ObservableCollection the Image get's updated. In cases if only the underlying Imagefile changes I want to have the Image Updated. I'm trying this by calling

            CachedImage.InvalidateCache(imagepath, FFImageLoading.Cache.CacheType.Disk);

But nothing is updated.

This is the Bound Object:

```
    [ImplementPropertyChanged]
    [JsonObject(IsReference = true)]
    public class CatalogueEntry : INotifyPropertyChanged
    {
        [JsonProperty(IsReference = true)] 
        private Catalogue catalogue;

        public CatalogueEntry(Catalogue parent)
        {
            catalogue = parent;
        }

        public string Name { get; set; }

        public string Description { get; set; }

        public string UUID { get; set; }

        public string ThumbPath
        {
            get { return catalogue.FolderPath + UUID + ".jpg"; }
        }

        public Catalogue Parent
        {
            get { return catalogue; }

            set { catalogue = value; }
        }

        public event PropertyChangedEventHandler PropertyChanged;
    }

```

In Xaml I bind ImageSource to ThumbPath.

What should I do?

**Comment by daniel-luberda** [comment]
After invalidation you should reapply the source

eg. 

```C#
var previous = ThumbPath;
ThumbPath = null;
ThumbPath = previous;
```

or call new `ReloadImage` method (not available on nuget yet).

**Comment by escamoteur** [comment]
ReloadImage would be great, but how do I access the Imageinstance inside a ListView when creating the View in Xaml.

As you saw ThumbPath has no setter. So reapplying the source won't work.
I know tried to manually call OnPropertyChanged,
This works, but only once, after that the Image is no longer updated.

**Comment by escamoteur** [comment]
I know used your Workaround reapplying the source again and it works only once, after that it does noch update the image anymore.
Ich I change the CachedImage to Image it works.

**Comment by daniel-luberda** [comment]
@escamoteur You should invalidate memory cache type.

**Comment by escamoteur** [comment]
Why? Its a FileImageresource.

**Comment by daniel-luberda** [comment]
Images for a given path are cached also in memory (if you use multiple images with same path, the only one bitmap instance is used).

**Comment by escamoteur** [comment]
I try:

            CachedImage.InvalidateCache(imagepath, FFImageLoading.Cache.CacheType.Memory);
           ((IngredientsMainPageModel) this.PreviousPageModel).UpdateLastItemImage();


And it updates only the first time, after that it always displays the same Image. with Image it works

**Comment by daniel-luberda** [comment]
Do you use any transformations or downsampling? There was a scenario when cache wasn't invalidated correctly with those features. See this: 

- https://github.com/molinch/FFImageLoading/commit/cac2927b7d19a1a69462caa6786bfcd098a8492c
- https://github.com/molinch/FFImageLoading/commit/1f110c264cfcd406403b103927706b46e799f4c9

I'll make a nuget with those features very soon.

**Comment by escamoteur** [comment]
I use Downsampling.Can I just pull from GitHub?

**Comment by daniel-luberda** [comment]
Sure. Just remember to use `removeSimilar = true` in `InvalidateCache` method.

**Comment by escamoteur** [comment]
what does it?

**Comment by daniel-luberda** [comment]
FFImageLoading caches bitmaps in memory. If you use downsampling or transformations, the same image can have different versions (600dp size, 100dp size, circleborder), those informations are added to cache key, eg: 

`image.png`
`image.png;600x0`
`image.png;100x0`
`image.png;CircleTransformation`

If you use `removeSimilar = true`, you'll invalidate all of them.

**Comment by escamoteur** [comment]
Now with removeSimilar = true the Image only flickers and does nbot change at all.
With removeSimilar = false it again changes only once.

**Comment by escamoteur** [comment]
I just git an Exception in

		public override void SetImageDrawable(Drawable drawable)
		{
			var previous = Drawable;

			base.SetImageDrawable(drawable);

			UpdateDrawableDisplayedState(drawable, true);
			UpdateDrawableDisplayedState(previous, false);
		}

System.ArgumentException: 'jobject' must not be IntPtr.Zero.

Parameter name: jobject

**Comment by daniel-luberda** [comment]
Can you make a repro? It really should work without problems, if you: 

- Invalidate cache
- Reload Image

> System.ArgumentException: 'jobject' must not be IntPtr.Zero.

I'll need more details.

**Comment by escamoteur** [comment]
This is the StackTrace:

System.ArgumentException'jobject' must not be IntPtr.Zero. Parameter name: jobject
Raw
Android.Runtime.JNIEnv.CallNonvirtualObjectMethod(IntPtr jobject, IntPtr jclass, IntPtr jmethod)JNIEnv.g.cs:586
Android.Widget.ImageView.get_Drawable()Android.Widget.ImageView.cs:542
FFImageLoading.Views.ManagedImageView.SetImageDrawable(Drawable drawable)
FFImageLoading.Forms.Droid.CachedImageRenderer.UpdateBitmap(CachedImage previous)
FFImageLoading.Forms.Droid.CachedImageRenderer.OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
<Module>.invoke_void_object_PropertyChangedEventArgs(object, PropertyChangedEventArgs)(wrapper delegate-invoke)
Xamarin.Forms.BindableObject.OnPropertyChanged(string propertyName)
Xamarin.Forms.BindableObject.SetValueActual(BindableProperty property, BindablePropertyContext context, object value, bool currentlyApplying, SetValueFlags attributes, bool silent)
Xamarin.Forms.BindableObject.SetValueCore(BindableProperty property, object value, SetValueFlags attributes, SetValuePrivateFlags privateAttributes)
Xamarin.Forms.BindingExpression.ApplyCore(object sourceObject, BindableObject target, BindableProperty property, bool fromTarget)
Xamarin.Forms.BindingExpression.Apply(bool fromTarget)
Xamarin.Forms.BindingExpression.BindingExpressionPart.<PropertyChanged>b__38_0()
Java.Lang.Thread.RunnableImplementor.Run()Thread.cs:36
Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this)Java.Lang.IRunnable.cs:71
at (wrapper dynamic-method) System.Object:61dfb1a0-98f3-44df-896f-08e2fb8886f5 (intptr,intp

**Comment by escamoteur** [comment]
Hi Daniel,
I think I wait for the next nuget-Package

**Comment by daniel-luberda** [comment]
@escamoteur 2.0.7-rc1 released.

**Comment by escamoteur** [comment]
Hi Daniel,

did you find something concerning my problem?

Best
Thomas

**Comment by daniel-luberda** [comment]
@escamoteur I reviewed some code parts but I didn't find anything. If I had a repo project I could test it further.

**Comment by escamoteur** [comment]
If it is no problem for you I could send you my current project, as it 
is still in the beginning.

------ Originalnachricht ------
Von: "Daniel Luberda" <notifications@github.com>
An: "molinch/FFImageLoading" <FFImageLoading@noreply.github.com>
Cc: "escamoteur" <thomas@burkharts.net>
Gesendet: 23.02.2016 18:26:13
Betreff: Re: [FFImageLoading] CachedImage.InvalidateCache has no effect 
(#147)

**Comment by daniel-luberda** [comment]
Sure, just give me some info where to look at :)If it's a bug, it would be nice to have it fixed before 2.0.7 stable release.

**Comment by escamoteur** [comment]
I just did some major refactoring, actually I'm preparing to bundle some 
part of my App as PCL for CRUD Dialogues and put it on GitHub soon.
As soon as everything works again ;-) I send you a donwload link.

------ Originalnachricht ------
Von: "Daniel Luberda" <notifications@github.com>
An: "molinch/FFImageLoading" <FFImageLoading@noreply.github.com>
Cc: "escamoteur" <thomas@burkharts.net>
Gesendet: 23.02.2016 18:29:46
Betreff: Re: [FFImageLoading] CachedImage.InvalidateCache has no effect 
(#147)

**Comment by daniel-luberda** [comment]
Sure :) Cheers!

**Comment by daniel-luberda** [comment]
Ok. Got it. It's because: https://github.com/molinch/FFImageLoading/blob/master/source/FFImageLoading.Shared/ImageService.cs#L309

As you see. Method is not awaited. :/ As a temporary solution, you could add a `Task.Delay` after invalidate call. We would need to convert that method to Async to fix that.

**Comment by escamoteur** [comment]
Hi Daniel,

yes a Task.Delay(300) did the job.

Best
Thomas

------ Originalnachricht ------
Von: "Daniel Luberda" <notifications@github.com>
An: "molinch/FFImageLoading" <FFImageLoading@noreply.github.com>
Cc: "escamoteur" <thomas@burkharts.net>
Gesendet: 23.02.2016 21:28:18
Betreff: Re: [FFImageLoading] CachedImage.InvalidateCache has no effect 
(#147)

**Comment by daniel-luberda** [comment]
It's fixed in RC2 nuget (use new methods)
